app/Model/Correlation.php

```diff
@@ -289,27 +289,45 @@ private function __getMatchingAttributes($value)
         // stupid hack to allow statically retrieving the constants
         ClassRegistry::init('Attribute');
         $conditions = [
-            'OR' => [
-                'Attribute.value1' => $value,
-                'AND' => [
-                    'Attribute.value2' => $value,
-                    'NOT' => ['Attribute.type' => Attribute::PRIMARY_ONLY_CORRELATING_TYPES]
-                ]
-            ],
             'NOT' => [
                 'Attribute.type' => Attribute::NON_CORRELATING_TYPES,
             ],
             'Attribute.disable_correlation' => 0,
             'Event.disable_correlation' => 0,
             'Attribute.deleted' => 0
         ];
-        $correlatingAttributes = $this->Attribute->find('all', [
-            'conditions' => $conditions,
-            'recursive' => -1,
-            'fields' => $this->getFieldRules(),
-            'contain' => $this->getContainRules(),
-            'order' => [],
-        ]);
+
+        $valueConditions = [
+            'Attribute.value1' => $value,
+            'AND' => [
+                'Attribute.value2' => $value,
+                'NOT' => ['Attribute.type' => Attribute::PRIMARY_ONLY_CORRELATING_TYPES]
+            ]
+        ];
+
+        $correlatingAttributes = [];
+
+        // Instead of relying upon SQL to optimize the query with sort_union, instead make
+        // up to three requests to get the correlations up to the correlation limit set.
+        // Not every MySQL setup will opt for sort_union, making correlations very slow on some setups.
+        foreach ($valueConditions as $valueKey => $valueTerm) {
+            $conditionsPart = $conditions;
+            $conditionsPart[$valueKey] = $valueTerm;
+
+            $correlatingAttributesPart = $this->Attribute->find('all', [
+                'conditions' => $conditionsPart,
+                'recursive' => -1,
+                'fields' => $this->getFieldRules(),
+                'contain' => $this->getContainRules(),
+                'order' => []
+            ]);
+
+            if (!empty($correlatingAttributesPart)) {
+                $correlatingAttributes = array_merge($correlatingAttributes, $correlatingAttributesPart);
```

Comment:
Please use 
```php
array_push($correlatingAttributes, ...$correlatingAttributesPart);
```
This will not create new array, but instead will append `$correlatingAttributesPart` to existing array `$correlatingAttributes` – so this is faster and consumes less memory.